sharktank/sharktank/layers/paged_attention.py

```diff
@@ -280,33 +349,47 @@ def read(
         page_tables = self.unflatten_page_tables(state)  # 6D
         page_table = page_tables[self.block_to_pipeline_lookup[transformer_block_index]]
 
-        bs, block_seq_len, *_ = page_ids.shape
-        # Blocks dim 1,2 according to the configured block stride.
-        blocked_shape = [
-            bs,
-            block_seq_len,
-            self.cache_partition_count,
-            self.block_seq_stride,
-            self.head_count_kv // self.shard_count,
-            self.attn_head_dim,
-        ]
-
-        # Gather both partitions and split post gather. This is more
-        # computationally efficient without gather fusion:
-        subblock_table = page_table.flatten(start_dim=0, end_dim=1)
-        page_stride = self.pipeline_to_block_count[
-            self.block_to_pipeline_lookup[transformer_block_index]
-        ]
-
-        transformer_block_index = torch.full(
-            (bs, block_seq_len), transformer_block_index, device=self.device
+        # TODO: IREE codegen cannot handle strided gathers properly today.
+        # vector.gather codegen for IREE is simply broken for strided gathers.
+        # This is sad, but to work around it, we flatten the strides into the
+        # gather indices. This is okay, because we aren't exactly linearizing
+        # two indices vectors into one, but only adding strides to it.
+        #
+        # The page table layout is organized as:
+        #   [page_id, attn_layer, cache_partition, page]
+        # Where the cache line can be 0 (k) or 1 (v).
+        #
+        # For a particular attention layer, attn_layer and cache_partition are
+        # fixed, and we are gathering over the page_id dimension. We flatten
+        # page_id, attn_layer, cache_partition into a single dimension, and
+        # gather over it, allowing us to bypass IREE's strided gather codegen.
+
+        # Get strided k/v page ids.
+        k_strided_page_ids = (
+            (page_ids * self.transformer_block_count * self.cache_partition_count)
+            + (transformer_block_index * self.cache_partition_count)
+            + 0
```

Comment:
Something looks wrong with the indexing. I have some fixes landing in `paged_attention.py` to try and cleanup the pipelining code. If you land the kernel work I can likely rework the changes in here.